### PR TITLE
Using IsNullOrEmpty in checks

### PR DESCRIPTION
**Pull request by Maxwe11**
Using IsNullOrEmpty in checks instead of direct checks on null or empty

### CODE REVIEW

**Comment by pdelvo** [comment]
I think this is a good change, but as in #391 `string` should be used instead of `String`.

**Comment by Maxwe11** [comment]
@pdelvo, thanks, fixed.

File: src/Common/src/System/Xml/XmlTextWriter.cs [hunk 1]
```diff
@@ -501,7 +501,7 @@ public override void WriteStartElement(string prefix, string localName, string n
                 }
                 else
                 {
-                    if ((ns != null && ns.Length != 0) || (prefix != null && prefix.Length != 0))
+                    if (!string.IsNullOrEmpty(ns) || !String.IsNullOrEmpty(prefix))
```
Comment (stephentoub) [inline src/Common/src/System/Xml/XmlTextWriter.cs]: Nit: the second "String" on this line still starts with a capital.

File: src/System.Xml.XPath.XmlDocument/src/System/Xml/XmlWellformedWriter.cs [hunk 2]
```diff
@@ -459,7 +459,7 @@ public override void WriteStartElement(string prefix, string localName, string n
                     {
                         ns = LookupNamespace(prefix);
                     }
-                    if (ns == null || (ns != null && ns.Length == 0))
+                    if (String.IsNullOrEmpty(ns))
```
Comment (stephentoub) [inline src/System.Xml.XPath.XmlDocument/src/System/Xml/XmlWellformedWriter.cs]: Nit: this one is also still capitalized

**Comment by stephentoub** [comment]
Other than the few straggler capitals, LGTM.

**Comment by mikedn** [comment]
I'd be a bit more careful with such changes, they may unnecessarily degrade performance.

**Comment by pdelvo** [comment]
@mikedn string.IsNullOrEmpty is most likely being inlined anyway so the resulting machine code should be exactly the same

**Comment by mikedn** [comment]
"Likely inlined" is not the same as "always inlined" and given the lack of performance regression testing I question the wisdom of making such changes. Especially in code that's 10 years old and not under active development.

**Comment by bbowyersmyth** [comment]
Micro-benchmarking the two techniques over 10 million iterations:
**x64**
IsNullOrEmpty is 2ms faster (*27ms vs 29ms*) than when length > 0 is used. Exactly the same when length != 0 is used.

**x86**
IsNullOrEmpty is 2ms slower (*30ms vs 28ms*) than the length checks.

Haven't checked the new compilers or mono, but small differences over a large run.

**Comment by shunsukeaida** [comment]
I'd rather be more curious about what benefit we'll get from using `IsNullOrEmpty()` than the performance (which I guess would be just faster or slower by a small margin). 

It certainly looks better than null check chore && arbitrary length check, but still.

File: src/Common/src/System/Xml/XmlConvertEx.cs [hunk 3]
```diff
@@ -66,7 +66,7 @@ public static string[] SplitString(string value)
 
         public static Uri ToUri(string s)
         {
-            if (s != null && s.Length > 0)
```
Comment (AdamSpeight2008) [inline src/Common/src/System/Xml/XmlConvertEx.cs]: Woud this `if( s?.Length = 0 ) return;` also work?

Comment (alexperovich) [inline src/Common/src/System/Xml/XmlConvertEx.cs]: it would have to be 'if ( s?.Length ?? 0 > 0) ;' because the return value of 's?.Length' is a nullable

Comment (mellinoe) [inline src/Common/src/System/Xml/XmlConvertEx.cs]: We aren't using C#6 features as of right now, but that looks like it should be equivalent.

**Comment by krwq** [comment]
Looks good to me :+1:

File: src/Common/src/System/Xml/XmlConvertEx.cs (hunk 3 above)
Comment (sharwell) [inline src/Common/src/System/Xml/XmlConvertEx.cs]: @alexperovich The following two expressions have identical behavior:

* `s?.Length > 0`
* `(s?.Length ?? 0) > 0`

The first comparison is the "lifted to null" version of the operation on two integers, and considers a `null` value unordered with respect to any value (i.e. `default(int?) > 0` is false and so is `default(int?) <= 0`).

**Pull request merged**